packages/neon-core/src/tx/transaction/main.ts

```diff
@@ -23,51 +25,87 @@ export function serializeArrayOf(prop: any[]): string {
   tx: Partial<TransactionLike> = {}
 ): Partial<TransactionLike> {
   const byte = ss.read();
-  return Object.assign(tx, { type: parseInt(byte, 16) });
+  const version = parseInt(byte, 16);
+  if (version !== 0) {
+    throw new Error(`Transaction version should be 0 not ${version}`);
+  }
+  return Object.assign(tx, { version });
 }
 
-export function deserializeVersion(
+export function deserializeNonce(
   ss: StringStream,
   tx: Partial<TransactionLike> = {}
 ): Partial<TransactionLike> {
-  const byte = ss.read();
-  return Object.assign({ version: parseInt(byte, 16) });
+  const nonce = parseInt(ss.read(4), 16);
+  return Object.assign(tx, { nonce });
 }
 
-export function deserializeAttributes(
+export function deserializeSender(
   ss: StringStream,
-  tx: Partial<TransactionLike>
+  tx: Partial<TransactionLike> = {}
 ): Partial<TransactionLike> {
-  const attributes = deserializeArrayOf(
-    TransactionAttribute.fromStream,
-    ss
-  ).map(i => i.export());
-  return Object.assign(tx, { attributes });
+  return Object.assign(tx, { sender: ss.read(20) });
 }
 
-export function deserializeInputs(
+export function deserializeScript(
+  ss: StringStream,
+  tx: Partial<TransactionLike> = {}
+): Partial<TransactionLike> {
+  const script = ss.readVarBytes();
+  if (script.length === 0) {
+    throw new Error("Script should not be vacant.");
+  }
+  return Object.assign(tx, { script });
+}
+
+export function deserializeFee(
+  ss: StringStream,
+  tx: Partial<TransactionLike> = {}
+): Partial<TransactionLike> {
+  const systemFee = fixed82num(ss.read(8));
+  const networkFee = fixed82num(ss.read(8));
+  return Object.assign(tx, { systemFee, networkFee });
+}
+
+export function deserializeValidUntilBlock(
   ss: StringStream,
   tx: Partial<TransactionLike>
 ): Partial<TransactionLike> {
-  const inputs = deserializeArrayOf(TransactionInput.fromStream, ss).map(i =>
-    i.export()
-  );
-  return Object.assign(tx, { inputs });
+  const validUntilBlock = parseInt(ss.read(4), 16);
+  return Object.assign(tx, { validUntilBlock });
 }
 
-export function deserializeOutputs(
+export function getCosignersFromAttributes(
+  attrs: TransactionAttribute[] | TransactionAttributeLike[]
+): string[] {
+  return attrs
+    .filter(attr => toTxAttrUsage(attr.usage) === TxAttrUsage.Cosigner)
+    .map(attr => attr.data);
+}
+
+export function deserializeAttributes(
   ss: StringStream,
   tx: Partial<TransactionLike>
 ): Partial<TransactionLike> {
-  const outputs = deserializeArrayOf(TransactionOutput.fromStream, ss).map(i =>
-    i.export()
-  );
-  return Object.assign(tx, { outputs });
+  const attributes = deserializeArrayOf(
+    TransactionAttribute.fromStream,
+    ss
+  ).map(i => i.export());
+  const cosigners = getCosignersFromAttributes(attributes);
+  if (
+    !cosigners.every(
+      cosigner =>
+        cosigners.indexOf(cosigner) === cosigners.lastIndexOf(cosigner)
+    )
+  ) {
+    throw new Error("Cosigner should not duplicate.");

```

Comment:
Same as above, if this is an deserialization issue, lets not throw here. We can do validation at another layer, perhaps when we are transforming the TransactionLike model into the Transaction object. Separating our concerns allows for reusability too.

Reply:
https://github.com/neo-project/neo/blob/2ef5d34cd88facbdfd13fd81edd8ad25012879b3/neo/Network/P2P/Payloads/Transaction.cs#L96

Reply:
But as you said, we could do validation in other parts. If you still think so considering the core code, I will do it.

Reply:
yeah i think its fine to not do validation at the deserialization step unless the data is badly malformed.

Reply:
same, remove the check.

Reply:
Removed